### ISSUE TITLE
YoloV5 IP camera RTSP or MJPG inference via SSH from remote PC?

### ISSUE THREAD

**Issue by cleverpal-ai** [ultralytics/yolov5#10003]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I'm tinkering with IP camera inference with YoloV5:

`python3 detect.py --weights yolov7-e6e.pt --source http://192.168.0.4/mjpg/video.mjpg`
 
and here's what I get when I initiate the command from a remote machine (in my case it should be a Home Automation server based on Home Assistant).

`YOLOR 🚀 v0.1-115-g072f76c torch 1.12.1+cu102 CUDA:0 (GeForce GTX 1650, 3903.875MB)

Fusing layers...
Model Summary: 792 layers, 151687420 parameters, 817020 gradients
 Convert model to Traced-model...
 traced_script_module saved!
 model is traced!

KiTTY X11 proxy: Unsupported authorisation protocol
qt.qpa.xcb: could not connect to display max-HP-Z240-SFF-Workstation:10.0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/usr/local/lib/python3.8/dist-packages/cv2/qt/plugins" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: xcb.

Aborted
`

I understand that it might be a kind of SSH issue, thus I googled what I could do and configured X11 port forwarding.
If I connect to my Ubuntu Server with YoloV5 deployed via RDP, I receive the same error when I run inference via Terminal.

On local PC with Python+YoloV5 deployed I have no issues, when I run same command and see realtime video window.
Tried with 
`view-img` on & off - makes no difference

Installed older OpenCV (incl contrib & headless) version as was recommended in some thread - no luck.
The only thing I can run successfully is static saved JPG inference, but it doesn't solve my problem.
Please advise how I can run inference of live stream on the server in headless mode.

Thanks in advance!

**Comment by glenn-jocher** [comment]
YOLOv5 uses cv2 under the hood for streaming, so before you do anything with YOLOv5 test your stream with a simple cv2 script.

**Comment by SuperMaximus1984** [comment]
> YOLOv5 uses cv2 under the hood for streaming, so before you do anything with YOLOv5 test your stream with a simple cv2 script.

Yes, I did. The above issue seems to be an OpenCV error. 
I have a local setup and it works like a charm except for the stream case.
So I wonder what I need to do in order to make OpenCV/YoloV5 run inference on a live stream on a system without GUI (no monitor). Digged all relative topics, installed all possible packages of Ubuntu with no luck of headless setup to work.

**Comment by SuperMaximus1984** [comment]
Is it generally possible to run video analysis let's say at 1 FPS without actual displaying QT-based video?
My task is to trigger an alarm if during live video analysis a particular object is detected. The script should be running by server & server should sent HTTP GET messages to a client.

**Comment by glenn-jocher** [comment]
@SuperMaximus1984 👋 Hello! Thanks for asking about **handling inference results**. YOLOv5 🚀 [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) models allow for simple model loading and inference in a pure python environment without using `detect.py`. 

### Simple Inference Example

This example loads a pretrained YOLOv5s model from PyTorch Hub as `model` and passes an image for inference. `'yolov5s'` is the YOLOv5 'small' model. For details on all available models please see the [README](https://github.com/ultralytics/yolov5#pretrained-checkpoints). Custom models can also be loaded, including custom trained PyTorch models and their [exported](https://docs.ultralytics.com/yolov5/tutorials/model_export) variants, i.e. ONNX, TensorRT, TensorFlow, OpenVINO YOLOv5 models.

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # yolov5n - yolov5x6 official model
#                                            'custom', 'path/to/best.pt')  # custom model

# Images
im = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, URL, PIL, OpenCV, numpy, list

# Inference
results = model(im)

# Results
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.
results.xyxy[0]  # im predictions (tensor)

results.pandas().xyxy[0]  # im predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie

results.pandas().xyxy[0].value_counts('name')  # class counts (pandas)
# person    2
# tie       1
```
<img width="550" src="https://user-images.githubusercontent.com/26833433/149679662-b2021c78-59b2-4f52-b768-1e5161537c85.png">

See [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) for details.

Good luck 🍀 and let us know if you have any other questions!

**Comment by LeSoleil1** [comment]
Hi,
I did this change which works for me:
In _general.py_:
```Python
def is_SSH() -> bool:
    """Check if the process runs through SSH."""
    if 'SSH_CONNECTION' in os.environ:
        return True
    return False
```

Also added assertion to `check_imshow` function in the same file (_general.py_):
```Python
def check_imshow(warn=False):
    # Check if environment supports image displays
    try:
        assert not is_notebook()
        assert not is_docker()
        assert not is_SSH() # ADDED 
        ...
```

**Comment by glenn-jocher** [comment]
@LeSoleil1 hello! Thank you for sharing your solution to the `imshow` error. While the code change you made seems to have resolved the issue for you, it's important to note that `imshow` requires a display to render the image and this may not always be possible in a headless environment, even with this change. In general, it's recommended that scripted analysis is done using headless libraries, such as OpenCV, which provide full-frame analysis of a video and can be run on a headless system. When needing to display the image, look for methods that can save the image to disk or send it to another system for display.

If you have any further questions or issues, please do not hesitate to let us know.

**Comment by pburma** [comment]
The webcam and http/rtsp streams are handled the same, so if you comment out the imshow section in this if statement you can skip the gui video loader. 

if webcam:
        view_img = check_imshow(warn=True)

If you go to line 226 in detect.py you can see how the view_img is used in that if block to run all the cv2 stuff which looks to me like what loads the gui display, so just keeping view_img False should do the trick which is part of the run() function params. 

view_img=False,

**Comment by glenn-jocher** [comment]
@pburma, that's correct! Setting `view_img=False` within the `run()` function parameters in `detect.py` will prevent the script from attempting to display the video stream using OpenCV's GUI functions, which is what you want for a headless environment.

By commenting out or bypassing the `view_img = check_imshow(warn=True)` line when `webcam` is `True`, you're ensuring that `view_img` remains `False`, thus avoiding any GUI-related errors when running the script over SSH or on a server without a display.

Just make sure that any other parts of your code that might attempt to display images or video frames are also disabled or handled in a way that is compatible with a headless setup.

If you encounter any further issues or have additional questions, feel free to reach out. Happy coding! 🚀
